### PR TITLE
Guide: Connect a Micronaut JMS Application to an AWS SQS Queue

### PR DESCRIPTION
**Pull request by sbodvanski**

### CODE REVIEW

File: guides/micronaut-jms-aws-sqs/groovy/src/main/groovy/example/micronaut/Application.groovy [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package example.micronaut
```
Comment (burtbeckwith) [inline guides/micronaut-jms-aws-sqs/groovy/src/main/groovy/example/micronaut/Application.groovy]: delete all 3 `Application` classes - these are generated with the rest of the initial app. You only need to add new or modified source files

File: guides/micronaut-jms-aws-sqs/java/src/main/java/example/micronaut/DemoProducer.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+import io.micronaut.jms.annotations.Queue;
+import io.micronaut.messaging.annotation.MessageBody;
+
+import static io.micronaut.jms.sqs.configuration.SqsConfiguration.CONNECTION_FACTORY_BEAN_NAME;
+
+@JMSProducer(CONNECTION_FACTORY_BEAN_NAME)
+public interface DemoProducer {
+
+    @Queue("demo_queue")
+    public void send(@MessageBody String body);
```
Comment (burtbeckwith) [inline guides/micronaut-jms-aws-sqs/java/src/main/java/example/micronaut/DemoProducer.java]: remove `public`

File: guides/micronaut-jms-aws-sqs/groovy/src/main/groovy/example/micronaut/DemoConsumer.groovy [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import io.micronaut.jms.annotations.Queue
+import io.micronaut.messaging.annotation.MessageBody
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+
+import static io.micronaut.jms.sqs.configuration.SqsConfiguration.CONNECTION_FACTORY_BEAN_NAME
+
+@JMSListener(CONNECTION_FACTORY_BEAN_NAME)
+class DemoConsumer {
+    private static Logger LOGGER = LoggerFactory.getLogger(DemoConsumer.class);
```
Comment (burtbeckwith) [inline guides/micronaut-jms-aws-sqs/groovy/src/main/groovy/example/micronaut/DemoConsumer.groovy]: remove this and the 2 imports and annotate the class with `@Slf4j('LOGGER')`

File: guides/micronaut-jms-aws-sqs/kotlin/src/main/kotlin/example/micronaut/SqsClientFactory.kt [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+import jakarta.inject.Singleton
+
+@Factory
+class SqsClientFactory {
+
+    @Singleton
+    fun sqsClient(environment: Environment?): AmazonSQS {
+        return AmazonSQSClientBuilder
+            .standard()
+            .withRegion(Regions.US_EAST_1)
```
Comment (burtbeckwith) [inline guides/micronaut-jms-aws-sqs/kotlin/src/main/kotlin/example/micronaut/SqsClientFactory.kt]: use a static import to be consistent with the Java and Groovy classes

File: guides/micronaut-jms-aws-sqs/kotlin/src/main/kotlin/example/micronaut/SqsClientFactory.kt [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+import io.micronaut.aws.sdk.v1.EnvironmentAWSCredentialsProvider
+import io.micronaut.context.annotation.Factory
+import io.micronaut.context.env.Environment
+import jakarta.inject.Singleton
+
+@Factory
+class SqsClientFactory {
+
+    @Singleton
+    fun sqsClient(environment: Environment?): AmazonSQS {
```
Comment (burtbeckwith) [inline guides/micronaut-jms-aws-sqs/kotlin/src/main/kotlin/example/micronaut/SqsClientFactory.kt]: remove `?`

File: guides/micronaut-jms-aws-sqs/kotlin/src/main/kotlin/example/micronaut/DemoProducer.kt [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+import io.micronaut.jms.annotations.JMSProducer
+import io.micronaut.jms.annotations.Queue
+import io.micronaut.jms.sqs.configuration.SqsConfiguration
+import io.micronaut.messaging.annotation.MessageBody
+
+@JMSProducer(SqsConfiguration.CONNECTION_FACTORY_BEAN_NAME)
+interface DemoProducer {
+
+    @Queue("demo_queue")
+    fun send(@MessageBody body: String?)
```
Comment (burtbeckwith) [inline guides/micronaut-jms-aws-sqs/kotlin/src/main/kotlin/example/micronaut/DemoProducer.kt]: remove `?` unless null body is allowed

File: guides/micronaut-jms-aws-sqs/kotlin/src/main/kotlin/example/micronaut/DemoConsumer.kt [hunk 7]
```diff
@@ -0,0 +1,21 @@
+package example.micronaut
+
+import io.micronaut.jms.annotations.JMSListener
+import io.micronaut.jms.sqs.configuration.SqsConfiguration
+import io.micronaut.messaging.annotation.MessageBody
+import com.example.DemoConsumer
```
Comment (burtbeckwith) [inline guides/micronaut-jms-aws-sqs/kotlin/src/main/kotlin/example/micronaut/DemoConsumer.kt]: remove this import

File: guides/micronaut-jms-aws-sqs/kotlin/src/main/kotlin/example/micronaut/DemoConsumer.kt [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+import io.micronaut.jms.annotations.Queue
+import org.slf4j.LoggerFactory
+
+@JMSListener(SqsConfiguration.CONNECTION_FACTORY_BEAN_NAME)
+class DemoConsumer {
+
+    @Queue(value = "demo_queue", concurrency = "1-3")
+    fun receive(@MessageBody body: String?) {
```
Comment (burtbeckwith) [inline guides/micronaut-jms-aws-sqs/kotlin/src/main/kotlin/example/micronaut/DemoConsumer.kt]: remove `?` unless null body is allowed

File: guides/micronaut-jms-aws-sqs/java/src/main/java/example/micronaut/DemoConsumer.java [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+import io.micronaut.jms.annotations.Queue;
+import io.micronaut.messaging.annotation.MessageBody;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static io.micronaut.jms.sqs.configuration.SqsConfiguration.CONNECTION_FACTORY_BEAN_NAME;
+
+@JMSListener(CONNECTION_FACTORY_BEAN_NAME)
+public class DemoConsumer {
+    private static Logger LOGGER = LoggerFactory.getLogger(DemoConsumer.class);
```
Comment (burtbeckwith) [inline guides/micronaut-jms-aws-sqs/java/src/main/java/example/micronaut/DemoConsumer.java]: should be `final`

File: guides/micronaut-jms-aws-sqs/groovy/src/main/groovy/example/micronaut/DemoConsumer.groovy [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+import io.micronaut.jms.annotations.JMSListener
+import io.micronaut.jms.annotations.Queue
+import io.micronaut.messaging.annotation.MessageBody
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+
+import static io.micronaut.jms.sqs.configuration.SqsConfiguration.CONNECTION_FACTORY_BEAN_NAME
+
+@JMSListener(CONNECTION_FACTORY_BEAN_NAME)
```
Comment (burtbeckwith) [inline guides/micronaut-jms-aws-sqs/groovy/src/main/groovy/example/micronaut/DemoConsumer.groovy]: If possible all Groovy classes should be annotated with `@CompileStatic`. I've had problems using this with controllers, but generally it works and will generate more efficient bytecode

File: guides/micronaut-jms-aws-sqs/kotlin/src/main/kotlin/example/micronaut/DemoProducer.kt [hunk 11]
```diff
@@ -0,0 +1,13 @@
+package example.micronaut
+
+import io.micronaut.jms.annotations.JMSProducer
+import io.micronaut.jms.annotations.Queue
+import io.micronaut.jms.sqs.configuration.SqsConfiguration
+import io.micronaut.messaging.annotation.MessageBody
+
+@JMSProducer(SqsConfiguration.CONNECTION_FACTORY_BEAN_NAME)
```
Comment (burtbeckwith) [inline guides/micronaut-jms-aws-sqs/kotlin/src/main/kotlin/example/micronaut/DemoProducer.kt]: use static import to be consistent

File: guides/micronaut-jms-aws-sqs/src/main/resources/logback.xml [hunk 12]
```diff
@@ -0,0 +1,15 @@
+<configuration>
```
Comment (burtbeckwith) [inline guides/micronaut-jms-aws-sqs/src/main/resources/logback.xml]: delete this file - it's the same as the file that's generated by the `create-app` step during the build

File: guides/micronaut-jms-aws-sqs/micronaut-jms-aws-sqs.adoc [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+** An IAM user with enough permissions to create and manage a queue instances in SQS.
+** The AWS CLI configured to use the IAM user above.
+
+common:completesolution.adoc[]
+
+common:create-app-features.adoc[]
+
+== Micronaut JMS SQS Dependency
+
+To use the Micronaut JMS SQS integration, add the following dependency:
```
Comment (burtbeckwith) [inline guides/micronaut-jms-aws-sqs/micronaut-jms-aws-sqs.adoc]: The dependency is added by the `jms-sqs` feature in the create-app command. Either remove this section, or change it to point out to the user that the feature adds this dependency. That's probably not needed though - they can see the effect of using the feature by clicking the diff link in the "If you have an existing Micronaut application ..." section

**Review by burtbeckwith — CHANGES_REQUESTED** [review]
looks great, only minor changes needed

File: guides/micronaut-jms-aws-sqs/metadata.json [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+  "title": "Connect a Micronaut JMS Application to an AWS SQS Queue",
+  "intro": "Learn how to connect JMS Application to an AWS SQS Queue",
+  "authors": ["Slavko Bodvanski"],
+  "tags": ["jms", "messaging", "aws"],
+  "categories": ["Messaging"],
+  "publicationDate": "2022-11-15",
+  "apps": [
+    {
+      "name": "default",
+      "features": ["jms-sqs"]
```
Comment (burtbeckwith) [inline guides/micronaut-jms-aws-sqs/metadata.json]: add `graalvm` feature

File: guides/micronaut-jms-aws-sqs/groovy/src/main/groovy/example/micronaut/Application.groovy (hunk 1 above)
Comment (sbodvanski) [reply guides/micronaut-jms-aws-sqs/groovy/src/main/groovy/example/micronaut/Application.groovy]: Thank you. That's good to know.

File: guides/micronaut-jms-aws-sqs/micronaut-jms-aws-sqs.adoc (hunk 13 above)
Comment (sbodvanski) [reply guides/micronaut-jms-aws-sqs/micronaut-jms-aws-sqs.adoc]: Good point

**Review by alvarosanchez — CHANGES_REQUESTED** [review]
This guide is using AWS SDK v1's `com.amazonaws.services.sqs.AmazonSQS`. It should also (or only) demonstrate how to use v2's `software.amazon.awssdk.services.sqs.SqsClient`, which is [preconfigured](https://micronaut-projects.github.io/micronaut-aws/latest/guide/#sqs) by our `micronaut-aws-sdk-v2`

**Comment by sbodvanski** [comment]
> AWS SDK v1's

I think that the current SQS sub-module in micronaut-jms supports only AWS SDK v1's (https://github.com/micronaut-projects/micronaut-jms/blob/master/jms-sqs/build.gradle). My understanding is that to support SDK V2, we will have to add a new module to micronaut-jms. Probably, that should be addressed in a separate ticket.

@alvarosanchez please take a look and let me know if I overlooked something, and if there is a way to work around the version restriction in the micronaut-jms module.

**Comment by alvarosanchez** [comment]
This is so funny, because a random guy already noticed that some time ago: https://github.com/micronaut-projects/micronaut-jms/issues/35

😅

**Comment by sbodvanski** [comment]
> micronaut-projects/micronaut-jms#35
:-D

**Comment by sbodvanski** [comment]
> This is so funny, because a random guy already noticed that some time ago: [micronaut-projects/micronaut-jms#35](https://github.com/micronaut-projects/micronaut-jms/issues/35)
> 
> 😅

Do you guys think it is still worth finishing this guide (JMS AWS SDK v1.x integration)? When support for SDK v2 is added to micronaut-jms, we could provide the V2 guide.

File: guides/micronaut-jms-aws-sqs/java/src/main/java/example/micronaut/SqsClientFactory.java [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+import com.amazonaws.services.sqs.AmazonSQS;
+import com.amazonaws.services.sqs.AmazonSQSClientBuilder;
+import io.micronaut.aws.sdk.v1.EnvironmentAWSCredentialsProvider;
+import io.micronaut.context.annotation.Factory;
+import io.micronaut.context.annotation.Requires;
+import io.micronaut.context.env.Environment;
+import jakarta.inject.Singleton;
+
+@Factory  // <1>
+@Requires(notEnv = Environment.TEST)
```
Comment (alvarosanchez) [inline guides/micronaut-jms-aws-sqs/java/src/main/java/example/micronaut/SqsClientFactory.java]: This will confuse readers, but actually you don't need it. Simply remove `@Requires` from here. Will add an inline comment in the test factory.

File: guides/micronaut-jms-aws-sqs/groovy/src/test/groovy/example/micronaut/MicronautguideSpec.groovy [hunk 16]
```diff
@@ -0,0 +1,9 @@
+package example.micronaut
+
+
+import spock.lang.Specification
+
+
+class MicronautguideSpec extends Specification {
+    // Localstack based tests are not supported.
```
Comment (alvarosanchez) [inline guides/micronaut-jms-aws-sqs/groovy/src/test/groovy/example/micronaut/MicronautguideSpec.groovy]: If Spock tests are not to be implemented, simply disable them in metadata.json, eg:

https://github.com/micronaut-projects/micronaut-guides/blob/master/guides/micronaut-aws-lambda-eventbridge-event/metadata.json#L9

File: guides/micronaut-jms-aws-sqs/java/src/test/java/example/micronaut/TestSqsClientFactory.java [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+import io.micronaut.aws.sdk.v1.EnvironmentAWSCredentialsProvider;
+import io.micronaut.context.annotation.Factory;
+import io.micronaut.context.annotation.Requires;
+import io.micronaut.context.env.Environment;
+import jakarta.inject.Singleton;
+
+import java.util.Optional;
+
+@Factory
+@Requires(env = Environment.TEST)
```
Comment (alvarosanchez) [inline guides/micronaut-jms-aws-sqs/java/src/test/java/example/micronaut/TestSqsClientFactory.java]: This is redundant since the class belong to the test sources, so no way it can be loaded outside tests.

Instead, add `@Replaces(SqsClientFactory.class)`

**Review by alvarosanchez — CHANGES_REQUESTED** [review]

**Comment by alvarosanchez** [comment]
> Do you guys think it is still worth finishing this guide (JMS AWS SDK v1.x integration)? When support for SDK v2 is added to micronaut-jms, we could provide the V2 guide.

Sure! Just address pending feedback, and then we can assign it to Sergio for publishing

**Comment by sbodvanski** [comment]
When I download the generated sources from the guide, I can compile them in IDE and run the app as well as the tests. However, the micronaut-guides CI build is failing because of:
```
Executing 'micronaut-jms-aws-sqs-maven-java' tests
Error:  COMPILATION ERROR : 
Error:  /home/runner/work/micronaut-guides/micronaut-guides/build/code/micronaut-jms-aws-sqs/micronaut-jms-aws-sqs-maven-java/src/main/java/example/micronaut/SqsClientFactory.java:[3,29] package com.amazonaws.regions does not exist
Error:  Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.10.1:compile (default-compile) on project micronautguide: Compilation failure
Error:  /home/runner/work/micronaut-guides/micronaut-guides/build/code/micronaut-jms-aws-sqs/micronaut-jms-aws-sqs-maven-java/src/main/java/example/micronaut/SqsClientFactory.java:[3,29] package com.amazonaws.regions does not exist
```

Not sure what I have missed here. The missing package referred to in the compilation error above is from the `com.amazonaws:aws-java-sdk-core` library.  I've found that it has been already added here: https://github.com/micronaut-projects/micronaut-guides/blob/master/buildSrc/src/main/resources/pom.xml#L21.

Do I need to introduce this lib or reintroduce as part of the new feature in the 'buildSrc/src/main/java/io/micronaut/guides/feature' ?

**Comment by alvarosanchez** [comment]
That pom.xml is only for having automatically upgradeable dependency versions, but it alone won't make it part of the generated build

**Comment by sbodvanski** [comment]
Have an additional question regarding the micronaut-guides  test sources compile task. Not sure why compile fails for Kotlin with the following error:

```
Error:  Failed to execute goal org.jetbrains.kotlin:kotlin-maven-plugin:1.6.10:test-compile (test-compile) on project micronautguide: Compilation failure: Compilation failure: 

Error:  /home/runner/work/micronaut-guides/micronaut-guides/build/code/micronaut-jms-aws-sqs/micronaut-jms-aws-sqs-maven-kotlin/src/test/kotlin/example/micronaut/MicronautguideTest.kt:[38,16] Unresolved reference: LocalStack
Error:  /home/runner/work/micronaut-guides/micronaut-guides/build/code/micronaut-jms-aws-sqs/micronaut-jms-aws-sqs-maven-kotlin/src/test/kotlin/example/micronaut/MicronautguideTest.kt:[43,9] Unresolved reference: LocalStack
```

What did I overlook here?

The same approach having `groovy` test source folder as well as `java` test source folder, which contains the LocalStack common helper class works for the Groovy task.

File: buildSrc/src/main/java/io/micronaut/guides/feature/Localstack.java [hunk 18]
```diff
@@ -26,6 +26,6 @@ public void processSelectedFeatures(FeatureContext featureContext) {
     @Override
     public void apply(GeneratorContext generatorContext) {
         super.apply(generatorContext);
-        addDependency(generatorContext, "aws-java-sdk-core", Scope.TEST);
+        addDependency(generatorContext, "aws-java-sdk-core");
```
Comment (alvarosanchez) [inline buildSrc/src/main/java/io/micronaut/guides/feature/Localstack.java:29]: Is this change intentional?

File: guides/micronaut-jms-aws-sqs/micronaut-jms-aws-sqs.adoc [hunk 19]
```diff
@@ -0,0 +1,133 @@
+common:header.adoc[]
+
+common:requirements.adoc[]
+* An https://aws.amazon.com/[AWS account] with:
+** An IAM user with enough permissions to create and manage a queue instances in SQS.
+** The AWS CLI configured to use the IAM user above.
+
```
Comment (alvarosanchez) [inline guides/micronaut-jms-aws-sqs/micronaut-jms-aws-sqs.adoc:7]:
```suggestion

common:aws.adoc[]

```

**Review by alvarosanchez — APPROVED** [review]

**Comment by alvarosanchez** [comment]
@sbodvanski please click on "Resolve conversation" for the feedback items that you already addressed, otherwise it's impossible for reviewers to tell what is done and what is still pending

**Comment by sbodvanski** [comment]
> @sbodvanski please click on "Resolve conversation" for the feedback items that you already addressed, otherwise it's impossible for reviewers to tell what is done and what is still pending

Will do. Thank you for pointing it out.

File: buildSrc/src/main/java/io/micronaut/guides/feature/Localstack.java (hunk 18 above)
Comment (sbodvanski) [reply buildSrc/src/main/java/io/micronaut/guides/feature/Localstack.java:29]: Yes, otherwise the ci build was failing with the compilation error:

```
Executing 'micronaut-jms-aws-sqs-maven-java' tests
Error:  COMPILATION ERROR : 
Error:  /home/runner/work/micronaut-guides/micronaut-guides/build/code/micronaut-jms-aws-sqs/micronaut-jms-aws-sqs-maven-java/src/main/java/example/micronaut/SqsClientFactory.java:[3,29] package com.amazonaws.regions does not exist
Error:  Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.10.1:compile (default-compile) on project micronautguide: Compilation failure
Error:  /home/runner/work/micronaut-guides/micronaut-guides/build/code/micronaut-jms-aws-sqs/micronaut-jms-aws-sqs-maven-java/src/main/java/example/micronaut/SqsClientFactory.java:[3,29] package com.amazonaws.regions does not exist 
```

**Review by alvarosanchez — APPROVED** [review]

**Review by graemerocher — APPROVED** [review]

**Review by burtbeckwith — APPROVED** [review]

File: guides/micronaut-jms-aws-sqs/micronaut-jms-aws-sqs.adoc [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+common:aws.adoc[]
+
+common:completesolution.adoc[]
+
+common:create-app-features.adoc[]
+
+== Create an application
+Let’s create a set of components that will use the Micronaut JMS to integrate with Amazon SQS. This will allow us to send and receive messages from the SQS. The SQS is a reliable, highly-scalable hosted queue for storing messages as they travel between applications or microservices. Amazon SQS moves data between distributed application components and helps you decouple these components.
```
Comment (sdelamo) [inline guides/micronaut-jms-aws-sqs/micronaut-jms-aws-sqs.adoc]: Missing link to SQS and use a blockquote for content which is quoting SQS documentation.

File: guides/micronaut-jms-aws-sqs/metadata.json [hunk 21]
```diff
@@ -0,0 +1,17 @@
+{
+  "title": "Connect a Micronaut JMS Application to an AWS SQS Queue",
+  "intro": "Learn how to connect JMS Application to an AWS SQS Queue",
+  "authors": ["Slavko Bodvanski"],
+  "tags": ["jms", "messaging", "aws"],
+  "categories": ["Messaging"],
+  "maximumJavaVersion": 16,
```
Comment (sdelamo) [inline guides/micronaut-jms-aws-sqs/metadata.json:7]: why fix maximumJavaVersion?

File: buildSrc/src/main/java/io/micronaut/guides/feature/Localstack.java (hunk 18 above)
Comment (sdelamo) [reply buildSrc/src/main/java/io/micronaut/guides/feature/Localstack.java:29]: This should not be necessary. This will add local stack to the main classpath.

File: guides/micronaut-jms-aws-sqs/java/src/main/java/example/micronaut/DemoController.java [hunk 22]
```diff
@@ -0,0 +1,20 @@
...
+public class DemoController {
+
+    private final DemoProducer demoProducer;
+
+    public DemoController(DemoProducer demoProducer) {  // <2>
+        this.demoProducer = demoProducer;
+    }
+
+    @Get("/demo")  // <3>
+    public void publishDemoMessages() {
```
Comment (sdelamo) [inline guides/micronaut-jms-aws-sqs/java/src/main/java/example/micronaut/DemoController.java:19]: The controller which has a side effect it publishes to a Queue. It should be mapped to POST instead of Get. 

I would annotate the method with `@Status(NO_CONENT)` and respond a 204 instead 200.

File: guides/micronaut-jms-aws-sqs/groovy/src/main/groovy/example/micronaut/DemoController.groovy [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+@Controller  // <1>
+class DemoController {
+
+    private final DemoProducer demoProducer
+
+    DemoController(DemoProducer demoProducer) {  // <2>
+        this.demoProducer = demoProducer
+    }
+
+    @Get("/demo")  // <3>
```
Comment (sdelamo) [inline guides/micronaut-jms-aws-sqs/groovy/src/main/groovy/example/micronaut/DemoController.groovy]: The controller which has a side effect it publishes to a Queue. It should be mapped to POST instead of Get. 

I would annotate the method with `@Status(NO_CONENT)` and respond a 204 instead 200.

File: guides/micronaut-jms-aws-sqs/kotlin/src/main/kotlin/example/micronaut/DemoController.kt [hunk 24]
```diff
@@ -0,0 +1,13 @@
+package example.micronaut
+
+import io.micronaut.http.annotation.Controller
+import io.micronaut.http.annotation.Get
+
+@Controller  // <1>
+class DemoController(private val demoProducer: DemoProducer) {  // <2>
+
+    @Get("/demo")  // <3>
```
Comment (sdelamo) [inline guides/micronaut-jms-aws-sqs/kotlin/src/main/kotlin/example/micronaut/DemoController.kt]: The controller which has a side effect it publishes to a Queue. It should be mapped to POST instead of Get. 

I would annotate the method with `@Status(NO_CONENT)` and respond a 204 instead 200.

File: guides/micronaut-jms-aws-sqs/groovy/src/test/groovy/example/micronaut/MicronautguideTest.groovy [hunk 25]
```diff
@@ -0,0 +1,46 @@
...
+import jakarta.inject.Inject;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestInstance;
+import org.testcontainers.junit.jupiter.Testcontainers;
+
+@Testcontainers(disabledWithoutDocker = true)
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+@MicronautTest
+class MicronautguideTest implements TestPropertyProvider {
```
Comment (sdelamo) [inline guides/micronaut-jms-aws-sqs/groovy/src/test/groovy/example/micronaut/MicronautguideTest.groovy]: why not write the groovy guide tests with Spock instead of Junit5 as we do for every guide?

File: guides/micronaut-jms-aws-sqs/metadata.json [hunk 26]
```diff
@@ -0,0 +1,17 @@
...
+  "testFramework": "junit",
```
Comment (sdelamo) [inline guides/micronaut-jms-aws-sqs/metadata.json]: it would be good to remove this and write Groovy tests with Spock.

File: guides/micronaut-jms-aws-sqs/groovy/src/test/groovy/example/micronaut/MicronautguideTest.groovy [hunk 27]
```diff
@@ -0,0 +1,46 @@
...
+    DemoController demoController
+
+    @Inject
+    DemoConsumer demoConsumer
+
+    @Test
+    void testItWorks() {
+        int messageCount = demoConsumer.getMessageCount()
+        assert messageCount == 0
+        demoController.publishDemoMessages()
```
Comment (sdelamo) [inline guides/micronaut-jms-aws-sqs/groovy/src/test/groovy/example/micronaut/MicronautguideTest.groovy]: I would prefer if we injected a HTTP Client and invoke the controller from the outside. 

Replace 

```
@Inject
DemoController demoController
```

with 

```
@Inject
@Client("/")
HttpClient httpClient
````

replace: 

`demoController.publishDemoMessages()`

with: 

`httpClient.toBlocking().exchange(HttpRequest.POST("/demo", Collections.emptyMap())`

File: guides/micronaut-jms-aws-sqs/java/src/test/java/example/micronaut/MicronautguideTest.java [hunk 28]
```diff
@@ -0,0 +1,49 @@
...
+import java.util.Map;
+
+
+@Testcontainers(disabledWithoutDocker = true)
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+@MicronautTest
+class MicronautguideTest implements TestPropertyProvider {
+
+    @Inject
+    DemoController demoController;
```
Comment (sdelamo) [inline guides/micronaut-jms-aws-sqs/java/src/test/java/example/micronaut/MicronautguideTest.java]: I would prefer if we injected a HTTP Client and invoke the controller from the outside. 

Replace 

```
@Inject
DemoController demoController
```

with 

```
@Inject
@Client("/")
HttpClient httpClient
````

replace: 

`demoController.publishDemoMessages()`

with: 

`httpClient.toBlocking().exchange(HttpRequest.POST("/demo", Collections.emptyMap())`

File: guides/micronaut-jms-aws-sqs/kotlin/src/test/kotlin/example/micronaut/MicronautguideTest.kt [hunk 29]
```diff
@@ -0,0 +1,46 @@
...
+import org.junit.jupiter.api.TestInstance
+import org.testcontainers.junit.jupiter.Testcontainers
+
+@Testcontainers(disabledWithoutDocker = true)
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+@MicronautTest
+internal class MicronautguideTest : TestPropertyProvider {
+
+    @Inject
+    lateinit var demoController: DemoController
```
Comment (sdelamo) [inline guides/micronaut-jms-aws-sqs/kotlin/src/test/kotlin/example/micronaut/MicronautguideTest.kt]: I would prefer if we injected a HTTP Client and invoke the controller from the outside. 

Replace 

```
@Inject
DemoController demoController
```

with 

```
@Inject
@Client("/")
HttpClient httpClient
````

replace: 

`demoController.publishDemoMessages()`

with: 

`httpClient.toBlocking().exchange(HttpRequest.POST("/demo", Collections.emptyMap())`

**Review by sdelamo — CHANGES_REQUESTED** [review]

File: guides/micronaut-jms-aws-sqs/micronaut-jms-aws-sqs.adoc (hunk 20 above)
Comment (sdelamo) [inline guides/micronaut-jms-aws-sqs/micronaut-jms-aws-sqs.adoc]:
```suggestion
Let’s create a set of components that will use the https://micronaut-projects.github.io/micronaut-jms/latest/guide/[Micronaut JMS] to send and receive messages from https://docs.aws.amazon.com/it_it/AWSSimpleQueueService/latest/APIReference/Welcome.html[AWS SQS]

____
Amazon SQS is a reliable, highly-scalable hosted queue for storing messages as they travel between applications or microservices. Amazon SQS moves data between distributed application components and helps you decouple these components.
____
```

**Review by sdelamo — CHANGES_REQUESTED** [review]

File: guides/micronaut-jms-aws-sqs/micronaut-jms-aws-sqs.adoc (hunk 20 above)
Comment (sbodvanski) [reply guides/micronaut-jms-aws-sqs/micronaut-jms-aws-sqs.adoc]: Added

File: guides/micronaut-jms-aws-sqs/metadata.json (hunk 21 above)
Comment (sbodvanski) [reply guides/micronaut-jms-aws-sqs/metadata.json:7]: Until we add a new module to micronaut-jms to support Java AWS SDK v2's, it seems that we are limited to Java version below 17:
https://github.com/aws/aws-sdk-java#maintenance-and-support-for-java-versions

File: guides/micronaut-jms-aws-sqs/java/src/main/java/example/micronaut/DemoController.java (hunk 22 above)
Comment (sbodvanski) [reply guides/micronaut-jms-aws-sqs/java/src/main/java/example/micronaut/DemoController.java:19]: Make sense. will change that.

File: guides/micronaut-jms-aws-sqs/groovy/src/test/groovy/example/micronaut/MicronautguideTest.groovy (hunk 25 above)
Comment (sbodvanski) [reply guides/micronaut-jms-aws-sqs/groovy/src/test/groovy/example/micronaut/MicronautguideTest.groovy]: Have replaced it with the Spock test implementation.

File: guides/micronaut-jms-aws-sqs/groovy/src/test/groovy/example/micronaut/MicronautguideTest.groovy (hunk 27 above)
Comment (sbodvanski) [reply guides/micronaut-jms-aws-sqs/groovy/src/test/groovy/example/micronaut/MicronautguideTest.groovy]: Will do it.

File: buildSrc/src/main/java/io/micronaut/guides/feature/Localstack.java (hunk 18 above)
Comment (sbodvanski) [reply buildSrc/src/main/java/io/micronaut/guides/feature/Localstack.java:29]: Actually, It does not add the localstack dependency to the main classpath, it just changes the scope of the `aws-java-sdk-core `dependency. There are no issues when the demo source is generated and imported to IDE. 
However, when I revert the change, the problem occurs during the micronut-guide project build. Tests cannot be compiled because `aws-java-sdk-core` artifacts are missing.

@sdelamo what do you suggest?
